만국박람회 [STEP 2] 민쏜, Judy

Judy-999 opened the pull request:
안녕하세요 하리보(@HARIBO) 민쏜과 Judy입니다 🤗  
만국박람회 Step 2 PR 보냅니다!
이번에도 잘 부탁드려요 😄

## 📱 실행 화면

| 메인화면 1 | 메인화면 2 |  
| -------- | -------- |
| ![](https://i.imgur.com/PTsQQdM.png) | ![](https://i.imgur.com/xsI1KI1.png)     |



| 한국의 출품작 화면 | 출품작 상세 화면 |
| -------- | -------- |
| ![](https://i.imgur.com/jXklHIW.png)| ![](https://i.imgur.com/thUc8JB.png)| 





## 질문드리고 싶은 점
### 1. tableView(_:cellForRowAt:) 메서드의 로직

셀의 Label에 대한 속성과 Accessory에 대한 속성을 코드로 설정하고 싶었습니다. 하지만 `ContentCell`은 커스텀 셀이라 테이블 뷰가 있는 `KoreanContentListViewController` 자체에서는 셀에 접근이 어려워 `ContentCell`에서 메서드를 통해 설정하고자 했습니다. 하지만 ` prepareForReuse` 함수를 사용하면 셀이 재사용될 때만 속성이 적용되었고, `init(style:reuseIdentifier:)` 함수는 원인을 파악할 수 없는 에러가 발생했습니다 🥲

```swift
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = contentTableView.dequeueReusableCell(withIdentifier: "contentCell", for: indexPath) as? ContentCell ?? ContentCell()
    let currentContent = contents[indexPath.row]
	
    // cell의 UI 요소의 attribute 설정
    cell.descriptionLabel.numberOfLines = 0
    cell.titleLabel.font = .preferredFont(forTextStyle: .title2)
    cell.accessoryType = .disclosureIndicator
    
    // cell의 내용 채우기
    cell.contentImageView.image = UIImage(named: currentContent.imageName)
    cell.titleLabel.text = currentContent.name
    cell.descriptionLabel.text = currentContent.shortDescription
        
    return cell
}
```

일단 현재는 `tableView(_:cellForRowAt:)`에서 모두 처리하도록 했습니다. 하지만 셀의 속성은 계속 변경할 필요 없이 한 번만 설정하면 되는데 `tableView(_:cellForRowAt:)` 함수가 불릴 때마다 불필요하게 코드가 실행되고, `tableView(_:cellForRowAt:)` 함수에서 셀의 내용과 속성을 동시에 설정하는 것이 적절하지 못하다고 생각했습니다. 그래서 혹시 다른 곳에서 처리 가능한지 알고 싶습니다! 아니면 이렇게 처리해도 괜찮은 방식일까요? 🤔

### 2. prepare 함수의 위치
segue로 화면이 이동하기 전에 작업을 수행하는 `prepare` 함수의 코드 상의 위치를 고민했습니다.

#### 1) ViewController 내부
새로운 파일을 생성할 때 Cocoa Touch Class에서 `UIViewController`를 선택하면 자동으로 ViewController 내부에 주석처리된 채로 `prepare` 함수가 존재해서 이 위치가 일반적이라고 생각했습니다. 

#### 2) 확장하여 UITableViewDelegate를 채택한 부분 (performSegue 위에)
아래처럼 `performSegue` 함수를 실행하여 실제로 `prepare` 함수를 호출하는 위치는 UITableViewDelegate의 `tableView(_: didSelectRowAt:)` 함수이기 때문에 동일한 위치에 두는 것이 좋다고 생각했습니다.
```swift
func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    performSegue(withIdentifier: "contentViewSegue", sender: indexPath)
}
```

따라서 현재 프로젝트에서는 UITableViewDelegate를 확장한 내부에 위치하도록 했는데 혹시 알맞은 위치인지, 만약 정답이 없다면 하리보의 생각은 어떠신지 궁금합니다!

### 3. segue의 네이밍
스토리보드에서 테이블 뷰인 한국의 출품작 화면에서 출품작 상세 화면으로 이동하는 segue를 지정했습니다. 해당 segue의 식별자를`ContentViewController`로 넘어간다는 의미로 **contentViewSegue**로 사용했습니다. 혹시 방향성을 나타내는 `"to"`나 다른 표현을 사용하는지 궁금했습니다. 일반적으로 사용하는 컨벤션이 있는지, 아니면 하리보는 보통 어떻게 작성하시는지 알고 싶습니다 😊

----
프로젝트를 진행하면서 어려웠거나 고민했던 점은 README에 자세하게 작성하였으니 읽어주시면 감사하겠습니다 😄
또한 저희가 미처 고려하지 못한 부분(예를 들어 구조/설계적인 측면 등)이 있다면 피드백 부탁드립니다!

Judy-999 committed:
feat: 메인화면 UI 구현

minsson committed:
feat: 메인화면 이미지 삽입

- ViewController 네이밍 MainViewController로 변경
- 메인화면 UI 추가 및 수정

minsson committed:
feat: decodeExpositionData 메서드 구현 및 관련 에러 타입 정의

- 메인화면에 표시할 Exposition 데이터를 관련 JSON 파일에서 파싱하는 메서드
- 에러처리에 필요한 DataHandlingError 구현

minsson committed:
feat: 메인 화면 내 모든 레이블을 채우는 setAllLabels 메서드 구현

- do-catch 문 내에서 setAllLabels 메서드 실행
- 레이블의 줄 수와 폰트 크기를 세팅하는 adjustLabelAttribute() 메서드 구현
- UI 변수 접근제어 수준 private으로 변경

Judy-999 committed:
fix: 메인화면에서 스크롤이 가능하도록 오토레이아웃 수정

Judy-999 committed:
feat: 방문객 수에 쉼표를 넣는 applyNumberFormat 메서드 구현

Judy-999 committed:
feat: 한국의 출품작 뷰 생성 및 네비게이션을 통한 화면 이동 구현

minsson committed:
feat: 한국의 출품작 화면의 테이블뷰 내부에서 사용될 ItemCell UI 구현

minsson committed:
feat: 한국의 출품작을 디코딩하는 메서드 구현

Judy-999 committed:
feat: 파싱한 한국의 출품작 데이터로 테이블 뷰의 내용을 표시

Judy-999 committed:
feat: 출품작 상세 페이지 UI 구현 및 파일 생성

minsson committed:
feat: 출품작별 상세페이지로 화면이동 및 데이터 전달 기능 구현

Judy-999 committed:
feat: 출품작별 상세페이지의 UI 요소를 채우는 메서드 구현

minsson committed:
refactor: 코드 전반에서 혼용되던 item과 content를 content로 통일

Judy-999 committed:
refactor: 출품작별 상세페이지로 데이터를 보낼 때 함수를 통해 보내도록 변경

minsson committed:
docs: README.md 내용 업데이트

Judy-999 committed:
docs: README.md 수정

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,80 @@
+
+import UIKit
+
+final class MainViewController: UIViewController {
```
`final` 과 접근지정자 사용을 잘해주고 계시네요 💯

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,80 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        adjustLabelAttribute()
+        setAllImages()
```
setAll--- 보다는 setup--- 으로 네이밍을 더 자주 사용합니다 🪄

minsson replied:
전체적으로 설정해준다는 의미에서 `setup`이라는 단어를 사용하는군요!! 수정했습니다 ☺️

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,80 @@
+    
+    private func setAllImages() {
+        expositionImageView.image = #imageLiteral(resourceName: "poster")
+        
```
exposition 이미지와 flag 이미지를 구분하는 용으로 공백을 주신것 같은데
이미지를 세팅 해주는 같은 역할을 하고 있어서 붙여주는게 코드 가독성에 더 좋을 것 같아요 😉

minsson replied:
넵, 그런 의도로 나눠놓았었는데, 세 줄 정도의 짧은 코드이므로 하리보 말씀대로 붙여주는 게 더 가독성이 좋을 것 같아요! 수정했습니다!

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,80 @@
+        rightFlagImageView.image = #imageLiteral(resourceName: "flag")
+    }
+    
+    private func adjustLabelAttribute() {
```
label 에 대한 처리는 setAllLabels() 에서 같이 해주는게 좋을될것 같아요!
setupLabels() 로 네이밍도 수정해주세요 👀

Judy-999 replied:
`setAllLabels` 메서드 명도 `setupLabels`로 변경하고,
`adjustLabelAttribute` 메서드 역시 label에 대한 처리이니 `setupLabels`로 통합했습니다!

HARlBO commented on Expo1900/Expo1900/Sources/Models/Error.swift:
```diff
@@ -0,0 +1,16 @@
+import Foundation
+import UIKit
```
UIKit 내부에 Foundation 프레임워크가 포함되어 있는걸 아시나요? 💡
이 파일에는 둘중 어떤 프레임워크를 import 하면 될까요?

minsson replied:
Apple 공식문서에 따르면, 각 프레임워크의 역할은 아래와 같아요.

`Foundation`: Access essential data types, collections, and operating-system services to define the base layer of functionality for your app. 

`UIKit`: Construct and manage a graphical, event-driven user interface for your iOS or tvOS app.

하리보의 코멘트를 읽은 후, UIKit 내부에 Foundation이 포함되어 있고, Foundation의 기능으로 해결이 된다면 Foundation을 넣으면 된다고 생각했어요.

그런데 다시 한번 생각해보니, 두 프레임워크의 역할을 정확하게 이해하지는 못했지만, 이 파일에서는 Foundation 프레임워크도 필요하지 않다는 생각이 들었어요.

따라서 두 줄 모두 지웠어요. 하리보는 어떻게 생각하시나요?

HARlBO replied:
UIKit 관련 코드가 필요 없는 경우에는 Foundation 만 import 하면 됩니다! 
주로 Model 관련 코드는 Foundation 만 필요한 경우가 많아서 드린 질문이였는데 이 파일에서는 두 프레임워크 모두 필요하지 않는게 맞네요 👏🏻  잘 접근하셨어요 🥇

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/KoreanContentListViewController.swift:
```diff
@@ -0,0 +1,72 @@
+}
+
+extension KoreanContentListViewController: UITableViewDelegate {
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
```
`prepare` 함수는 `UITableViewDelegate` 에 속한 함수가 아니기 때문에 요 위치보다는 
KoreanContentListViewController 클래스 내부로 옮겨주시면 됩니다 😉

Judy-999 replied:
넵 `KoreanContentListViewController` 클래스 내부로 옯겼습니다!

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/KoreanContentListViewController.swift:
```diff
@@ -0,0 +1,72 @@
+        let cell = contentTableView.dequeueReusableCell(withIdentifier: "contentCell", for: indexPath) as? ContentCell ?? ContentCell()
+        let currentContent = contents[indexPath.row]
+        
+        cell.descriptionLabel.numberOfLines = 0
```
요기 cell 에 대한 처리들은 ContentCell 내부에서 함수로 처리해주고
여기서 그 함수를 호출해 주면, 역할을 분리 할 수 있고 코드 가독성에도 좋습니다 ✨

minsson replied:
이런 방향은 생각해보지 못했네요! 😲

ContentCell 클래스에 아래 코드블럭과 같은 메서드를 만들고,
기존의 tableView(_:cellForRowAt:)에서는  `cell.setup(content: contents[indexPath.row])`라고 써서 호출해주었어요. 하리보가 말씀하신 내용을 잘 이해한 걸까요?

```swift
    func setup(content: Content) {
        descriptionLabel.numberOfLines = 0
        titleLabel.font = .preferredFont(forTextStyle: .title2)
        accessoryType = .disclosureIndicator
        
        contentImageView.image = UIImage(named: content.imageName)
        titleLabel.text = content.name
        descriptionLabel.text = content.shortDescription
    }
```

HARlBO replied:
네넵! 너무 잘 이해해 주셨어요 💯

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,25 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        descriptionLabel.numberOfLines = 0
```
요 라인도 `setUIComponents()` 함수로 옮겨주는게 어떨까요!🧐

Judy-999 replied:
Label에 대한 변경이고, Label도 UIComponent이니 `setUIComponents()` 메서드 안에 위치하는게 적절하는 것 같아 변경했습니다!

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,25 @@
+        setUIComponents()
+    }
+
+    func receiveContentData(_ data: Content) {
```
데이터를 receive 받는다 의 느낌보다는 세팅해주는 의미로 `setupContentData` 정도도 좋을 것 같아요!

minsson replied:
받아서 데이터를 세팅하는 게 목표이므로, 말씀해주신 것처럼 `setupContentData`라는 네이밍이 더 좋을 것 같아요! 변경했습니다!

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/ContentViewController.swift:
```diff
@@ -0,0 +1,25 @@
+import UIKit
+
+final class ContentViewController: UIViewController {
```
Content 라는 단어가 많은 의미를 가지고 있어서 여기서는 
Content 들의 상세 페이지인 느낌을 추가하기 위해서 `ContentDetailViewController` 같은 네이밍으로 해주면 더 명확하게 어떤 클래스인지 알수 있을것 같아요 😉

minsson replied:
`ContentViewController`라는 이름을 떠올리면서 조금 허전하다 싶었는데, `Detail`이라는 단어를 추가하니 더 명확하게 느껴지네요! 수정했습니다 ☺️

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/KoreanContentListViewController.swift:
```diff
@@ -0,0 +1,72 @@
+        contentTableView.delegate = self
+        contentTableView.rowHeight = 150
+
+        do {
```
데이터를 decode 하고, contents 를 세팅해주고 있는 요 do catch 문이 
에러 처리 등으로 더 길어 질수 있고, viewDidLoad 를 봤을 때
어떤 역할을 하고 있는지 바로 알기 위해 함수로 분리하면 좋을 것 같아요!
데이터베이스나 어딘가에 저장된 데이터를 가져올때는 주로 fetch--- 라는 네이밍을 많이 사용해요 🪄

Judy-999 replied:
넵 do-catch를 처리하는 구문이 두 곳에 있어서
`MainViewController`에서는 `fetchExpositionData` 메서드로 분리하고, `KoreanContentListViewController`에서는 `fetchContentData` 메서드로 분리했습니다!

HARlBO commented on Expo1900/Expo1900/Sources/Models/ContentCell.swift:
```diff
@@ -0,0 +1,15 @@
+    @IBOutlet weak var titleLabel: UILabel!
+    @IBOutlet weak var descriptionLabel: UILabel!
+    
+    override func awakeFromNib() {
```
super 호출외에 사용하지 않는 함수들은 지워주는게 좋을 것 같아요 👀

minsson replied:
지웠습니다! ☺️

HARlBO reviewed (commented):
안녕하세요 **민쏜, Judy** 😊
STEP2 도 고생 많으셨어요 👏🏻👏🏻👏🏻

일단 질문 주신 내용들에 대한 답변 먼저 드릴게요 💡 

****1. tableView(_:cellForRowAt:) 메서드의 로직****

`prepareForReuse` 함수는 셀이 재사용 될때 속성이 적용 되는 함수가 맞아요!
셀이 재사용 될때 간혹 이전에 사용한 셀에 있던 label 이나 image 같은 컴포넌트들이 초기화 되지 않고 그대로 보여지는 경우가 있어서, `prepareForReuse` 함수에서 미리 초기화 시켜주는 코드를 주로 넣어줍니다!

이럴 때는, KoreanContentListViewController 에서 `prepare` 함수에서 
`contentViewController.receiveContentData(contents[indexPath.row])` 해주신 것처럼 동일하게
ContentCell 내부에도 setup(content: Content) 함수를 만들어 주고, 그 함수 내부에서 ``cellForRowAt`` 에서 해주고 있는 cell 에 대한 세팅들을 해주고,
`cellForRowAt` 에서는 그 함수를 호출해주면 될것 같아요! 

****2. prepare 함수의 위치****

`prepare` 함수 위치는 아래 코멘트에도 드렸지만 ViewController 내부에 두시면 됩니다!
extension 으로 분리한 경우 호출 위치와는 상관 없이
UITableViewDelegate 에 해당하는 코드만 같이 묶어주시는게 좋습니닷 😉

****3. segue의 네이밍****

현재 네이밍 해주신것 처럼 **`contentViewSegue` 를 주로 사용해요!**


그리고 호옥시! 🧐 ViewController 간 데이터 전달하는 다른 방법들도 알고 계신가요?
`prepare` 함수를 통해 데이터 전달을 하는 방법을 선택하신 이유도 궁금합니다! 👀

나머지 내용들은 코멘트 확인해주세요 🙇🏻‍♀️

minsson committed:
refactor: 가독성 개선을 위한 메서드명 변경 및 메서드 내 개행 제거

- setAllImages() 메서드를 setupImages()로 변경
- 내부 코드는 모두 이미지를 세팅하는 같은 역할을 수행하므로, 개행 제거

minsson committed:
Merge branch 'ic_6_minsson' of https://github.com/minsson/ios-exposition-universelle into ic_6_minsson

Judy-999 committed:
refactor: 메서드 명 변경 및 Label 변경하는 메서드 하나로 합치기

- setAllLabels에서 setupLabels로 변경

minsson committed:
refactor: 불필요한 프레임워크 및 메서드 제거

minsson committed:
Merge branch 'ic_6_minsson' of https://github.com/minsson/ios-exposition-universelle into ic_6_minsson

Judy-999 committed:
refactor: prepare 함수의 위치 변경

- extension 내부에서 KoreanContentListViewController 내부로 이동

minsson committed:
refactor: cell의 속성 및 값을 부여하는 코드를 별도의 메서드로 분리

- tableView(_:cellForRowAt:)의 역할 일부를 ContentCell의 setup(content:) 메서드로 분리
- 메서드 역할을 분리하고 가독성을 개선하기 위함

Judy-999 committed:
refactor: Label 속성을 setUIComponents 메서드에서 처리하도록 변경

minsson committed:
refactor: 의미를 명확하게 하기 위한 클래스 및 메서드 네이밍 변경

- ContentViewController: 기존 Content라는 단어가 많은 의미를 갖고 있으므로, Content들의 상세 페이지인 의미를 전달하고자 ContentDetailViewController로 변경
- receiveContentData(_:): 데이터를 세팅해준다는 의미를 전달하고자 setupContentData(_:)로 변경

Judy-999 committed:
refactor: do-catch 구문을 메서드로 분리

- 에러처리가 길어질 수 있고, viewDidLoad에서 역할을 분리하기 위해 메서드로 분리
- 데이터를 가져오는 메서드이기 때문에 fetchContentData, fetchExpositionData로 명명

minsson commented:
하리보, 안녕하세요! 코멘트 감사합니다! ☺️
피드백 주신 내용은 수정하고, 저희의 생각도 남겨두었어요.

> 그리고 호옥시! 🧐 ViewController 간 데이터 전달하는 다른 방법들도 알고 계신가요?
prepare 함수를 통해 데이터 전달을 하는 방법을 선택하신 이유도 궁금합니다! 

라고 말씀하신 것과 관련해서는, 답변으로 저희가 README.md에 적어놓았던 내용을 가져와 봤어요! 우선 저희가 알고 있었던 방식과 장단점, 그리고 지금의 방식을 선택한 근거는 아래와 같아요. 혹시 더 좋은 방법이 있을까요?

### 출품작 리스트 화면에서 출품작별 페이지로의 화면이동 
#### 방식 1) prepare(for:sender:)만 사용 
- 장점: `prepare(for:sender:)` 함수만 사용해서 간편함
- 단점: 데이터를 전달하기 위해 파라미터인 sender에서 indexPath값을 추출해야 함
#### 방식 2) performSegue(withIdentifier:sender:) 사용
- 장점: sender로 indexPath값을 보내 전달할 데이터를 쉽게 얻을 수 있음
- 단점: `prepare(for:sender:)`와 `tableView(_: didSelectRowAt:)` 모두 사용해야 함
#### 방식 3) instantiateViewController(withIdentifier:) 사용
- 장점: `tableView(_: didSelectRowAt:)`에서만 처리할 수 있음
- 단점: segue 없이 코드로만 되어 있어 직관적이지 못함 
<br>

 **프로젝트 내에서는 두 번째 방식인 performSegue, prepare 사용** <br>
근거: segue를 이용해 화면 이동을 구현하는 것이 가장 직관적으로 느껴졌음. 테이블 뷰의 델리게이트 함수인 `tableView(_: didSelectRowAt:)` 함수 역시 사용해보는 것이 좋다고 생각함.


---

# 1. 피드백 해주신 부분과 관련된 질문들

## 1) UI 설정 코드와 값을 할당하는 코드의 역할 분리
피드백해주신 내용을 바탕으로, 현재 `setupLabels(of:)`, `setupImages()`, `setup(content:)`와 같은 함수를 사용하고 있어요.

이 세 함수는 모두 아래 코드와 같이 윗 부분은 각 UI 요소들의 속성을 설정해주고, 아래 부분에서는 값을 할당한다는 공통점이 있어요. 그걸 구분하기 위해 개행을 하고 있고요. 
```swift
    func setup(content: Content) {
        descriptionLabel.numberOfLines = 0
        titleLabel.font = .preferredFont(forTextStyle: .title2)
        accessoryType = .disclosureIndicator
        
        contentImageView.image = UIImage(named: content.imageName)
        titleLabel.text = content.name
        descriptionLabel.text = content.shortDescription
    }
``` 

⁉️ 그런데 저희 생각에는 위쪽과 아래쪽이 다른 역할을 한다고 느껴지고, 그래서 메서드의 역할 분리 측면에서 별도의 메서드로 묶는 게 나을 수 있겠다는 생각이 들어요. 하리보는 어떻게 생각하시나요?


## 2) ContentCell 내부의 setup(content: Content) 함수의 파라미터 네임

종합 코멘트에서 조언해주신대로 ContentCell 내부에도 setup(content: Content) 함수를 만들어봤어요. 

`cell.setup(content: contents[indexPath.row])`와 같이 사용하고 있는데요,
content가 두 번 반복되는 게 조금 마음에 걸려요. 

⁉️`contents[indexPath.row]`라는 아규먼트를 받아 setup 한다는 의미에서 `cell.setup(with: contents[indexPath.row])` 라는 네이밍은 어떨까요? 하리보는 어떻게 생각하시나요?

---

# 2. Step 2 추가 질문

아래 내용은 Step 2를 진행하면서 README.md에 작성했던 고민사항인데, 추가로 하리보의 조언을 얻고 싶어요!

## 1) catch 에러처리
우선 디코딩하는 동안 발생할 수 있는 에러를 처리하기 위해 DataHandlingError 타입을 구현해서 사용했습니다. DataHandlingError과 관련된 에러는 지정한 description을 출력하도록 처리했습니다. 추가적으로 그 외 모든 에러를 처리하기 위한 마지막 catch 블럭이 필요한데, 추가적으로 처리해야할 작업을 판단하기 어려웠습니다. 검색을 통해 아래 코드와 같이 처리할 수 있다는 점을 알게되어 적용했습니다!

```swift
do {
    contents = try decodeContentData()
} catch let error as DataHandlingError {
    print(error.description)
} catch {
    print("Unexpected error: \(error)") 
}
```

⁉️ 그런데 단순히 어떤 에러인지 출력하는 코드라면, 이러한 처리 방식이 어떤 것을 의도한 것인지 의문이 생겼습니다. 현재 프로젝트에서는 별도로 처리할 동작이 없어 이런 처리 방식도 유효하지만, 현업에서는 어떻게 처리하는지 알고 싶습니다!

##  2) 메인화면의 setAllImages 메서드에서 이미지를 불러오는 방법
#### #imageLiteral 사용
- 장점: 어떤 이미지가 들어가게 될지 코드 상에 시각적으로 보여 직관적
- 단점: 이미지가 동적으로 변경되어야 하는 경우 사용할 수 없음
#### UIImage(named:) 사용
- 장점: 이미지를 동적으로 변경할 수 있음
- 단점: 코드 상에서는 이미지 파일명을 통해 어떤 이미지일지 파악해야 함

프로젝트 내에서는 #imageLiteral을 사용
근거: 현재 관련 이미지가 정적으로 고정된 상태이므로, 코드 내에서 이미지를 시각적으로 볼 수 있는 #imageLiteral이 더 유용하다고 생각해서 사용했습니다.

⁉️ 현업에서는 어떤 방식이 더 자주 사용되는지, 저희가 생각해본 장단점이 있긴하지만 보통 두 가지 방법 중에서 어떤 기준을 두고 선택하는지 궁금합니다.

HARlBO commented on Expo1900/Expo1900/Sources/Controllers/ContentDetailViewController.swift:
```diff
@@ -0,0 +1,27 @@
+        guard let content = content else { return }
+        
+        descriptionLabel.numberOfLines = 0
+        
```
위 아래 라인 모두 descriptionLabel 세팅에 관련된 값이라 개행을 없애주면 좋을것 같아요!

minsson replied:
넵!

`descriptionLabel`은 같은 컴포넌트라 묶어주고, 아래의 `contentImageView.image = UIImage(named: content.imageName)`와 `title = content.name`은 별개의 내용이라 개행해주었어요.

```swift
private func setUIComponents() {
        guard let content = content else { return }
        
        descriptionLabel.numberOfLines = 0
        descriptionLabel.text = content.description
        
        contentImageView.image = UIImage(named: content.imageName)
        title = content.name
    }
```

다만, 아래 두 줄도 별개의 내용이므로 그 사이에도 개행이 필요하지 않을까 싶은데, 하리보가 작성해주신 코드 예시를 보면 이런 내용은 개행을 안 하시는 것 같더라고요. 

⁉️ 혹시 한 컴포넌트에 대한 내용이 두 줄 이상일 때만 개행하여 묶어주고, 한 줄인 컴포넌트들은 개행 없이 묶어주는 건가요?

HARlBO reviewed (approved):
코멘트 반영해주신 내용 확인했습니다 👍🏻 


화면 이동하는 방법을 README 에 잘 정리해 두셨었네요!

선택하신 방식은 하나의 스토리보드 내에서 ViewController 간의 segue 를 연결 해두었을 때 가능한 방식이고,
`instantiateViewController(withIdentifier:)` 방식은 다른 스토리 보드에 있더라도, 코드로 해당 스토리보드와 ViewController 를 identifier 를 통해 생성해서 접근할 수가 있는 방식이라는 점이 큰 차이일것 같아요!
이번 프로젝트는 화면이 많지 않고 하나의 스토리보드 내에서 작업이 가능해서 선택하신 방법도 좋은 방법이예요.

나중에 협업을 하다 보면 스토리보드 conflict 를 방지하기 위해 여러개의 스토리보드를 부고 화면 작업을 하는 경우가 많아서 저 포인트를 기억해두셨으면 해서 질문 드렸습니다 😉

---

****1. 피드백 해주신 부분과 관련된 질문들****

**1) UI 설정 코드와 값을 할당하는 코드의 역할 분리**

음 UI 관련 코드에 대한 규칙은 컨벤션을 정하기 나름이겠지만,
저는 보통 회사에서도 UI 관련 코드는 속성 설정과 값 할당을 나누지 않고 있어요.
현재는 UI 속성 설정, 값 할당 이렇게 나누어 주셨는데 저는 주로

```Swift
    func setup(content: Content) {
        descriptionLabel.numberOfLines = 0
         descriptionLabel.text = content.shortDescription

        titleLabel.font = .preferredFont(forTextStyle: .title2)
        titleLabel.text = content.name

        accessoryType = .disclosureIndicator
        contentImageView.image = UIImage(named: content.imageName)
    }
```

이런식으로 각각 같은 컴포넌트에 해당된 코드로 묶어주는 편이예요!

말씀해주신 속성 설정, 값 할당을 나누는 경우는
예를들어 ViewController 가 호출 될 때, 딱 한번만 설정해주면 되는 속성 관련 코드는 setup 함수 내에서 처리하고,
사용자 액션이나 어떤 이벤트에 따라 여러번 호출되면서 컴포넌트들이 변경되는 경우는 update 함수 내에서 처리하는 경우에는 적절할것 같습니닷!

****2) ContentCell 내부의 setup(content: Content) 함수의 파라미터 네임****

`cell.setup(with: contents[indexPath.row])`

네이밍 너무 좋습니다 💯 content 가 반복 되는게 마음에 걸리셨다니..! 벌써 Swift 네이밍 습관이 몸에 베이셨나봐요 👏🏻

---

****2. Step 2 추가 질문****

****1) catch 에러처리****

구현해주신 do - catch 문을 보면
파일을 불러오다 에러가 나면 DataHandlingError 를 통해 invalidFile 에러를 출력해 주고 있고, 
또 decodeExpositionData() 함수를 보게되면 

`try jsonDecoder.decode(Exposition.self, from: mainPage.data)`
에서 throw 해주는 예외들도 있겠되겠죠!

질문 주신 `print("Undexpected error: \(error)")` 는 
decode 하다가 에러가 발생하면 그 에러를 출력해 주는 역할을 하게 될거예요.

decode 하다가 발생할수 있는 에러들은 키 값이 잘못되었거나, 타입이 다르거나, 키 값이 존재하지 않은 등 엄청 다양하게 있는데요! 해당 에러를 출력해주어서 개발자가 어떤 에러인지 인지하고 바로 수정할 수 있도록 해줍니다!

****2) 메인화면의 setAllImages 메서드에서 이미지를 불러오는 방법****

#imageLiteral 에 적어주신 장점에 하나 더 추가하자면, 직접 asset 에 있는 이미지를 선택하기 때문에, UIImage(named:) 에서 이미지 명에 오타가 나거나 잘못되어 불러오지 못하는 경우를 방지 할 수 있어요!

하지만 asset 에 이미지들이 엄청 많아지게 되면 그만큼 이미지 찾는데도 오래걸리기도 해요!
음 현업에서는 imageLiteral을 사용하는 회사도 있었지만 주로 UIImage(named:)를 사용하는것 같아요!


코멘트 드린 사항은 다음 스텝에 반영해주시면 됩니다 😉 

STEP3 에서 만나요 🙋🏻‍♀️

minsson committed:
refactor: 가독성 개선을 위한 개행 수정

- descriptionLabel은 같은 컴포넌트에 대한 코드이므로 개행으로 분리함

minsson committed:
Merge branch 'ic_6_minsson' of https://github.com/minsson/ios-exposition-universelle into ic_6_minsson

Judy-999 committed:
fix: 한국의 출품작 페이지에서 설명이 생략되는 문제 수정

- 테이블 뷰의 rowHeight을 automaticDimension으로 설정해서 셀의 높이가 유동적으로 변하게 함
- 셀의 오토레이아웃 변경
